Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Exceptions.Add.cs

```diff
@@ -0,0 +1,236 @@
+
+            var expectedGroupPostDependencyException =
+                new GroupPostDependencyException(failedGroupPostStorageException);
+
+            this.storageBrokerMock.Setup(broker =>
+                broker.InsertGroupPostAsync(It.IsAny<GroupPost>()))
+                    .ThrowsAsync(sqlException);
+
+            // when
+            ValueTask<GroupPost> addGroupPostTask =
+                this.groupPostService.AddGroupPostAsync(someGroupPost);
+
+            // then
+            await Assert.ThrowsAsync<GroupPostDependencyException>(() =>
+                addGroupPostTask.AsTask());
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.InsertGroupPostAsync(It.IsAny<GroupPost>()),
+                    Times.Once);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogCritical(It.Is(SameExceptionAs(
+                    expectedGroupPostDependencyException))),
+                        Times.Once);
+
+            this.storageBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task ShouldThrowDependencyValidationExceptionOnAddIfGroupPostAlreadyExistsAndLogItAsync()
+        {
+            // given
+            GroupPost someGroupPost = CreateRandomGroupPost();
+            string randomMessage = GetRandomMessage();
+
+            var duplicateKeyException =
+                new DuplicateKeyException(randomMessage);
+
+            var alreadyExistsGroupPostException =
+                new AlreadyExistsGroupPostException(duplicateKeyException);
+
+            var expectedGroupPostDependencyValidationException =
+                new GroupPostDependencyValidationException(alreadyExistsGroupPostException);
+
+            this.storageBrokerMock.Setup(broker =>
+                broker.InsertGroupPostAsync(It.IsAny<GroupPost>()))
+                    .ThrowsAsync(duplicateKeyException);
+
+            // when
+            ValueTask<GroupPost> addGroupPostTask =
+                this.groupPostService.AddGroupPostAsync(someGroupPost);
+
+            GroupPostDependencyValidationException actualGroupPostDependencyValidationException =
+                await Assert.ThrowsAsync<GroupPostDependencyValidationException>(
+                    addGroupPostTask.AsTask);
+
+            // then
+            actualGroupPostDependencyValidationException.Should().BeEquivalentTo(
+                expectedGroupPostDependencyValidationException);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(
+                    expectedGroupPostDependencyValidationException))),
+                        Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.InsertGroupPostAsync(someGroupPost),
+                    Times.Once);
+
+            this.loggingBrokerMock.VerifyNoOtherCalls();
```

Comment:
There is one rule when verifying dependencies. You order these verification statements on the order by which they are called. 
In this case you first call `storageBroker` that throws exception and then `loggingBroker` logs the error. The order of these statements should be the same. This is the correct order: 

``` cs
this.storageBrokerMock.Verify(broker =>
    broker.InsertGroupPostAsync(someGroupPost),
        Times.Once);

this.loggingBrokerMock.Verify(broker =>
    broker.LogError(It.Is(SameExceptionAs(
            expectedGroupPostDependencyValidationException))),
                Times.Once);

this.storageBrokerMock.VerifyNoOtherCalls();
this.loggingBrokerMock.VerifyNoOtherCalls();

```

Reply:
Okay, correcting according to this way.